src/os2datascanner/utils/optimiser/optimiser.py

```diff
@@ -0,0 +1,314 @@
+    In our case, we just want to check if two containers
+    have the same components. This function detects any
+    elements that do not occur in both. If none are found,
+    they are identical."""
+    base = arr1
+    for k in arr2:
+        try:
+            base.remove(k)
+        except ValueError:
+            # Can't remove because it isn't there
+            return False
+    # Return not (is there anything left in base). True if empty else False
+    return not base
+
+def find_duplicates(arr1, arr2):
+    # Yields any element that appears in both.
+    for k in arr1:
+        if k in arr2:
+            yield k
+
+def inverse_extend(arr1, deletions):
+    """Given a base array and a list of elements,
+    removes from the base array each 
+    element in that list."""
+    for k in deletions:
+        try:
+            arr1.remove(k)
+        except ValueError:
+            raise ValueError("ValueError occurred in inverse_extend()")
+    return arr1
+
+def load_json(path, extensive=False):
+    # Loads json object from given path
+    with open(path) as file:
+        obj = file.read()
+        if extensive:
+            for id in rule_ids:
+                obj = obj.replace(rule_ids[id], str(id))
+        return json.loads(obj)
+ 
+def dump_json(obj, output_path):
+    # Dumps python dict object as json to given path
+    with open(output_path, "wt") as file:
+        json.dump(obj, file, indent=2)
+
+def display(data):
+    # Displays CustomContainer object as a dict
+    try:
+        print(json.dumps(data, indent=2))
+    except:
+        print(json.dumps(json.loads(data), indent=2))
+
+def arr_switch(arr:list, deletion, insertion):
+    """Maybe useless, but quite frequently had
+    to remove a and append b to a list, so
+    this function does just that.
+    Switches in given array a for b."""
+    arr.remove(deletion)
+    arr.append(insertion)
+    return arr
+
+def remove_duplicates(arr):
+    """Traditional set() doesn't work with
+    custom classes (CustomContainer in 
+    this case), and removes it entirely."""
+    mem = []
+    for k in arr:
+        if k in mem:
+            pass
+        else:
+            mem.append(k)
+            yield k
+
+def container_count_op(container:CustomContainer, op):
+    # Returns amount of containers *with given operator* directly in another one.
+    c = 0
+    for e in container.components:
+        if isinstance(e, CustomContainer):
+            if e.operator == op:
+                c += 1
+    return c
+
+def contains_container(cont):
+    # Boolean of (cont has any containers directly inside ?)
+    return any(isinstance(i, CustomContainer) for i in cont.components)
+
+DIR = "/home/magenta/osdatascanner/src/os2datascanner/utils/optimiser/"
+IN_PATH = "/home/magenta/osdatascanner/src/os2datascanner/utils/optimiser/original_rule.json"
+OUT_PATH = "/home/magenta/osdatascanner/src/os2datascanner/utils/optimiser/output_rule.json"
```

Comment:
I understand that this is a prototype, but I think these three lines make it into a prototype that can only work on your computer. :-)

Reply:
Yes, i forgot to remove these